### PR TITLE
Nodeset offline for ppc64

### PR DESCRIPTION
**Pull request by gurevichmark**
This pull request fixes problem reported by issue #450.
The perl modules for handling nodeset offline for ppc64 did not remove boot configuration files in /tftpboot.
The perl modules for handling x86 were not changed. They appear to correctly remove the boot configuration files.

This pull request does the following:
1. Updates man page for **nodeset** command to remove the `iscsiboot` option
2. Removes duplicate `getstate` function definition from **petitboot.pm**
3. Removes appropriate boot configuration files from /tftpboot directory when **nodeset \<node\> offline** is issued for ppc64le virtual and ppc64le non-virtual
4. Removes dhcp entries from dhcp.leases file for nodes specified in **nodeset \<node\> offline**
5. Changes the status returned by **nodeset \<node\> stat** from `boot` to `offline` if no boot configuration file is found.

Unit test was performed against ppc64le VM and ppc64le non-VM systems. Single machine and multiples.

### CODE REVIEW

File: xCAT-server/lib/xcat/plugins/petitboot.pm [hunk 1]
```diff
@@ -535,6 +542,14 @@ sub process_request {
           }
       }
   }
+
```
Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/petitboot.pm]: Please use 4 spaces as indent

File: xCAT-server/lib/xcat/plugins/petitboot.pm [hunk 2]
```diff
@@ -188,8 +192,8 @@ sub setstate {
     #        arg=>['-s','filename = \"xcat/nonexistant_file_to_intentionally_break_netboot_for_localboot_to_work\";']},$callback);
     #print $pcfg "bye\n";
     close($pcfg);
-  } elsif ($kern and $kern->{kernel}) {
-    #It's time to set yaboot for this node to boot the kernel..
+  } elsif ($kern and $kern->{kernel} and $cref and $cref->{currstate} ne "offline") {
+    #It's time to set yaboot for this node to boot the kernel, but only if not offline directive
```
Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/petitboot.pm]: this comment is only for your change, the ``yaboot`` in comment message should be replaced to ``petitboot``

File: xCAT-server/lib/xcat/plugins/grub2.pm [hunk 3]
```diff
@@ -713,6 +724,17 @@ sub process_request {
         }
     }
 
+    if ($args[0] eq 'offline') {
+        # If nodeset directive was offline we need to remove the architecture file link and remove dhcp entries
+        foreach my $osimage (keys %osimagenodehash) {
```
Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/grub2.pm]: for performance consideration, we should put all the nodes in an array and then call makedhcp -d at once.

File: xCAT-server/lib/xcat/plugins/petitboot.pm [hunk 4]
```diff
@@ -535,6 +542,14 @@ sub process_request {
           }
       }
   }
+
+  if ($args[0] eq 'offline') {
+    # If nodeset directive was offline we need to remove dhcp entries
+    foreach my $node (@normalnodeset) {
+      $sub_req->({ command => ['makedhcp'],arg=>['-d'],
```
Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/petitboot.pm]: for performance consideration, we should put all the nodes in an array and then call makedhcp -d at once.

Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/petitboot.pm]: do we need to run ``makedhcp -d`` for netboot=xnba?

Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/petitboot.pm]: A common question, does the 'nodeset <node> offline' can make the the <node> get into the discovery in the next boot?

Comment (gurevichmark) [inline xCAT-server/lib/xcat/plugins/petitboot.pm]: @daniceexi I think we should add `makedhcp -d` to netboot=xnba. But I did not want to change code that was not specific for the issue #450.
If you agree, I can add code to call `makedhcp -d` to xnba.pm.
Should I:
a. Use the same pull request as this code for ppc64 ?
b. Use a separate pull request and reference the same issue #450 
b. Open a separate issue for xnba and fix it under a separate pull request ?

Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/petitboot.pm]: @gurevichmark I'd like to make changes for xnba in this pull request to ease to track. But should be in different commits.

Since we don't need to run discovery for ppc64, I prefer to leave ppc64 as is.

Comment (gurevichmark) [inline xCAT-server/lib/xcat/plugins/petitboot.pm]: @daniceexi Ok. I will make a similar change for **xnba**. What about **pxe** ? Should I make a similar change there ?
Also, can you please explain what you meant by: *Since we don't need to run discovery for ppc64, I prefer to leave ppc64 as is.* ? Did you mean that **yaboot** or **grub2** does not need to be changed ?

**Comment by daniceexi** [comment]
@gurevichmark sorry to give the incorrect information. I meant that we don't need to take care of removing dhcp entry from dhcpd.leases for ppc64. But I don't think it was correct, please ignore this message.

I think we should make all the netboot methods have the consistent behaviour for ``nodeset offline``:
* remove configuration file
* remove bootloader file for the node if necessary
* remove dhcp entry from dhcpd.lease

**Comment by daniceexi** [comment]
Looks like we cannot finish this pull request in 2.12.1, so move it to 2.12.2

**Comment by gurevichmark** [comment]
Second commit added to fix code review suggestions:
1. Indentation
2. Single call to `makedhcp -d` with a list of nodes instead of one by one.

File: xCAT-server/lib/xcat/plugins/grub2.pm [hunk 5]
```diff
@@ -724,15 +724,16 @@ sub process_request {
         }
     }
 
+    my @makedhcp_nodes;
```
Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/grub2.pm]: could you put this variable ``@makedhcp_nodes`` inside the L728 block?

I suggest the variable name could be 'rmdhcp_nodes'

**Comment by gurevichmark** [comment]
@daniceexi Thank you for you suggestion. I checked in another commit with the suggested changes for `grub2.pm` and also simplified code for `xnba.pm`

**Comment by daniceexi** [comment]
I have reviewed the change and agree to merge. Thanks @gurevichmark

**Comment by gurevichmark** [comment]
@daniceexi Thank you for reviewing the code changes.

* Should `pxe.pm` also be changed to remove dhcp entries from `dhcpd.lease` when `nodeset offline` is issued ?
* Should I fix the formatting to `xnba.pm`, `grub2.pm` and `petitboot.pm` using `perltidy` ?

**Comment by daniceexi** [comment]
I don't think there's too much people are still using netboot=pxe. So we don't need to change it.

Yes, please run perltidy against these files and make a new new pull request.

In fact, I am thinking to run perltidy against all the xCAT files in the this release beginning phase.
